### ISSUE TITLE
Emission of NullabilityAttribute(2) on unconstrained generic type parameters causes incorrect runtime reporting of nullability in .NET 9

### ISSUE THREAD

**Issue by mlorbetske** [dotnet/roslyn#74837]
**Version Used**: 9.0.100-preview.7.24407.12

**Steps to Reproduce**:

```csharp
using System.Reflection;

var parameter = typeof(HasGenericParameter<object>).GetConstructors()[0].GetParameters()[0];
var nullabilityContextInfo = new NullabilityInfoContext().Create(parameter);

Console.WriteLine("HasGenericParameter<object>.ctor(object parameter)");
Console.WriteLine(nullabilityContextInfo.ReadState);
Console.WriteLine(nullabilityContextInfo.WriteState);

parameter = typeof(HasGenericParameter<object?>).GetConstructors()[0].GetParameters()[0];
nullabilityContextInfo = new NullabilityInfoContext().Create(parameter);

Console.WriteLine("HasGenericParameter<object?>.ctor(object? parameter)");
Console.WriteLine(nullabilityContextInfo.ReadState);
Console.WriteLine(nullabilityContextInfo.WriteState);

/*
.NET 8 Output
=============
HasGenericParameter<object>.ctor(object parameter)
NotNull
NotNull
HasGenericParameter<object?>.ctor(object? parameter)
NotNull
NotNull
*/

/*
.NET 9 Output
=============
HasGenericParameter<object>.ctor(object parameter)
Nullable
Nullable
HasGenericParameter<object?>.ctor(object? parameter)
Nullable
Nullable
*/
public class HasGenericParameter<T>(T parameter)
{
    public override string? ToString()
    {
        return parameter?.ToString();
    }
}
```

https://github.com/dotnet/runtime/issues/106636

**Expected Behavior**:

From the above example, it would be expected that the call to get the nullability of the parameter `HasGenericParameter<object>` would be `NotNull`, or at least `Unknown`, matching the CS8625 warning reported in the IDE when attempting to pass null to the constructor parameter.

**Actual Behavior**:

From the above example, the parameter is reported as nullable at runtime (because of the application of `Nullability(2)` to the unconstrained generic argument).

**Comment by jaredpar** [comment]
> From the above example, it would be expected that the call to get the nullability of the parameter HasGenericParameter<object> would be NotNull, or at least Unknown, 

This is a statement of the libraries representation of nullability which is a step removed from the compiler. What do you think is the incorrect behavior in the compiler here?

**Comment by jaredpar** [comment]
Had some time to dig through the associated runtime issue this morning. The compiler is not marking them as nullable here. The compiler is marking them as [being absent a specific set of constraints](https://github.com/dotnet/roslyn/blob/main/docs/features/nullable-metadata.md#type-parameters)

> No notnull, class, struct, unmanaged, or type constraints in #nullable enable context (equivalent to an object? constraint): NullableAttribute(2)

The current encoding accurately reflects the source declaration. How the library team interprets that, given the ambiguity of the situation, is a decision for them. There isn't a way for us to change the encoding here to help them out. Our encoding must accurately reflect the way the parameter is declarred so that it can be correctly round tripped through metadata.
